source/fuzz/transformation_add_relaxed_decoration.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_add_relaxed_decoration.h"
+
+#include "source/fuzz/fuzzer_util.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationAddRelaxedDecoration::
+TransformationAddRelaxedDecoration(
+    const spvtools::fuzz::protobufs::
+    TransformationAddRelaxedDecoration& message)
+    : message_(message) {}
+
+TransformationAddRelaxedDecoration::
+TransformationAddRelaxedDecoration(uint32_t result_id) {
+  message_.set_result_id(result_id);
+}
+
+bool TransformationAddRelaxedDecoration::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& transformation_context) const {
+  // |message_.result_id| must be the id of an instruction.
+  auto instr = ir_context->get_def_use_mgr()->GetDef(message_.result_id());
+  if (!instr) {
+    return false;
+  }
+  opt::BasicBlock* cur_block =
+      ir_context->get_instr_block(instr);
+  // current block must be a dead block
+  if (!(transformation_context.GetFactManager()->BlockIsDead(
+      cur_block->id()))) {
+    return false;
+  }
+  // The instruction must be numeric.
+  return IsNumeric(instr->opcode());
+}
+
+void TransformationAddRelaxedDecoration::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // Add a RelaxedPrecision decoration targeting |message_.result_id|.
+  ir_context->get_decoration_mgr()->AddDecoration(message_.result_id(),
+                                                  SpvDecorationRelaxedPrecision);
+}
+
+protobufs::Transformation TransformationAddRelaxedDecoration::ToMessage()
+const {
+  protobufs::Transformation result;
+  *result.mutable_add_relaxed_decoration() = message_;
+  return result;
+}
+
+bool TransformationAddRelaxedDecoration::IsNumeric(uint32_t opcode) {
+  switch (opcode) {
+    case SpvOpConvertFToU:
+    case SpvOpConvertFToS:
+    case SpvOpConvertSToF:
+    case SpvOpConvertUToF:
+    case SpvOpUConvert:
+    case SpvOpSConvert:
+    case SpvOpFConvert:
```

Comment:
I don't think we want this function because I don't think RelaxedPrecision is only applied to numeric instructions. But if we did want this function, I think we should move the original function to fuzzer_util.h (or elsewhere) and re-use it, instead of duplicating it.

Reply:
Actually, I think we do want this function. But we could maybe avoid the duplication.

Reply:
@alan-baker The validator doesn't seem to complain if I apply RelaxedPrecision to e.g. OpConstantTrue, or the result of a load.

Is it legitimate to apply RelaxedPrecision to ids that are non-numeric, with the expectation that it would just be ignored?

Reply:
I think the validator doesn't check what it is applied to. I agree it doesn't really make sense to apply to a bool.